xrft/xrft.py

```diff
@@ -116,50 +116,52 @@ def detrend_wrap(detrend_func):
     Wrapper function for `xrft.detrendn`.
     """
     def func(a, axes=None):
-        if len(axes) > 3:
-            raise ValueError("Detrending is only supported up to "
-                            "3 dimensions.")
-        if axes is None:
-            axes = tuple(range(a.ndim))
-        else:
-            if len(set(axes)) < len(axes):
-                raise ValueError("Duplicate axes are not allowed.")
 
-        for each_axis in axes:
-            if len(a.chunks[each_axis]) != 1:
-                raise ValueError('The axis along the detrending is upon '
-                                'cannot be chunked.')
+        if len(set(axes)) < len(axes):
+            raise ValueError("Duplicate axes are not allowed.")
 
-        if len(axes) == 1:
-            return dsar.map_blocks(sps.detrend, a, axis=axes[0],
-                                   chunks=a.chunks, dtype=a.dtype
-                                  )
-        else:
-            for each_axis in range(a.ndim):
-                if each_axis not in axes:
-                    if len(a.chunks[each_axis]) != a.shape[each_axis]:
-                        raise ValueError("The axes other than ones to detrend "
-                                        "over should have a chunk length of 1.")
-            return dsar.map_blocks(detrend_func, a, axes,
+        return dsar.map_blocks(detrend_func, a, axes,
                                    chunks=a.chunks, dtype=a.dtype
                                   )
 
     return func
 
-def _apply_detrend(da, axis_num):
+def apply_detrend(da, dim, axis_num, detrend_type):
     """Wrapper function for applying detrending"""
-    if da.chunks:
-        func = detrend_wrap(detrendn)
-        da = xr.DataArray(func(da.data, axes=axis_num),
+
+    if detrend_type not in ['constant','linear',None]:
+        raise NotImplementedError("%s is not a valid detrending option. Valid "
+                                  "options are: 'constant','linear', or None."
+                                  % detrend_type)
+
+    if detrend_type == 'constant':
+        return da - da.mean(dim=dim)
+
+    elif detrend_type == 'linear':
+        if len(dim) == 1:
+            p = da.polyfit(dim=dim[0], deg=1)
+            fit = xr.polyval(da[dim[0]], p.polyfit_coefficients)
+            return da - fit
+
+        elif len(dim) > 3:
+            raise ValueError("Detrending is only supported up to "
+                            "3 dimensions.")
+
+        # For > 1d detrend, need all non-FFT axes to have chunks of length 1
+        for d in da.dims:
+            if d not in dim:
+                da = da.chunk({d:1})
```

Comment:
```suggestion
                raise ValueError(f"Cannot apply detrending over dimension {d} because it is chunked. "
                                 f"Please rechunk your data first by calling `da.chunk({{d}: 1)`. ")
```

I would definitely not recommend silently and automatically rechunking the data. This could lead to big problems for large datasets. Instead, raise an error and ask the user to explicitly rechunk the data.

We would also need to make sure this scenario is covered by a test (e.g. `pytest.raises`).

Reply:
Sounds good. The current version of the code does this chunking, and to be honest I don't fully understand why it's needed (beyond the fact that it's for the 2d and 3d detrending via matrix algebra) so left it in. But I agree, we should raise an error instead of doing automatically.

Reply:
Clarification: these lines here actually chunk the data into chunks of length 1, so the error should state that dimensions not being detrended over must be chunked. I deleted a warning in the last PR that I will add back in to this effect.